src/services/elasticsearch/elasticsearch.service.ts

```diff
@@ -0,0 +1,166 @@
+      infer: true,
+    });
+    this.eventIndex = eventsConfig.whiteboard_event_index;
+    this.eventLoggingMode = eventsConfig.mode;
+
+    this.sendDataThrottled = throttle(
+      this._sendWhiteboardChangeEvent.bind(this),
+      eventsConfig.interval,
+    );
+  }
+
+  public sendWhiteboardChangeEvent(
+    roomId: string,
+    createdBy: string,
+    changes: DetectedChanges<ExcalidrawElement>,
+  ) {
+    if (this.eventLoggingMode === WhiteboardEventLoggingMode.none) {
+      return;
+    }
+
+    const baseDocument: BaseChangeEventDocument = {
+      '@timestamp': new Date(),
+      whiteboardId: roomId,
+      createdBy,
+      types: this.eventType(changes),
+    };
+
+    if (this.eventLoggingMode === WhiteboardEventLoggingMode.lite) {
+      this.dataToSend.push(baseDocument);
+    }
+
+    if (this.eventLoggingMode === WhiteboardEventLoggingMode.full) {
+      this.dataToSend.push({
+        ...baseDocument,
+        ...changes,
+      });
+    }
+
+    this.sendDataThrottled();
+  }
+
+  private async _sendWhiteboardChangeEvent() {
+    const result = await this.ingestBulk(this.dataToSend, this.eventIndex);
+
+    this.dataToSend.length = 0;
+
+    return result;
+  }
+
+  private async ingestBulk(data: unknown[], index: string): Promise<void> {
+    if (!this.elasticClient) {
+      return;
+    }
+
+    if (!data.length) {
+      return;
+    }
+
+    const operations = data.flatMap((doc) => [
+      { create: { _index: index } },
+      doc,
+    ]);
+
+    const bulkResponse = await this.elasticClient.bulk({ operations });
+
+    if (bulkResponse.errors) {
+      const erroredDocuments: ErroredDocument[] = [];
+      // The items array has the same order of the dataset we just indexed.
+      // The presence of the `error` key indicates that the operation
+      // that we did for the document has failed.
+      bulkResponse.items.forEach((action, i) => {
+        const operation = Object.keys(action)[0] as keyof typeof action;
+        if (action[operation]?.error) {
+          erroredDocuments.push({
+            // If the status is 429 it means that you can retry the document,
+            // otherwise it's very likely a mapping error, and you should
+            // fix the document before to try it again.
+            status: action[operation]?.status,
+            error: action[operation]?.error,
+            operation: operations[i * 2],
+            document: operations[i * 2 + 1],
+          });
+        }
+      });
+      this.logger.error(
+        `[${index}] - ${erroredDocuments.length} documents errored. ${
+          data.length - erroredDocuments.length
+        } documents indexed.`,
+      );
+    } else {
+      this.logger.verbose?.(`[${index}] - ${data.length} documents indexed`);
+    }
+  }
+
+  private eventType(changes: DetectedChanges<any>): DetectedChangesType[] {
+    const types: DetectedChangesType[] = [];
+
+    if (changes.inserted) {
+      types.push('insert');
```

Comment:
It will be better if you have a type / enum that has `inser` `update` `delete`, not pass a string directly.

Reply:
it's a typed string, which does not allow to use random strings - switched to const enum